cmd/charm-ingest/main.go

```diff
@@ -0,0 +1,179 @@
+be transferred.
+
+For example, the following whitelist specifies that the latest stable revision
+of the wordpress charm should be transferred, and revision 254 of the
+canonical-kubernetes bundle with all its charm dependencies should
+be made available in the beta and stable channels.
+
+	wordpress
+	cs:bundle/canonical-kubernetes-254 beta stable
+
+By default, entities will be copied from the global charm store (https://api.jujucharms.com/charmstore);
+this can be overridden by setting the JUJU_CHARMSTORE environment variable.
+
+The destination argument holds the URL of the charm store to copy charms into.
+The auth flag can be used to specify the admin username and password for the destination;
+if not specified, the user will be authenticated with the Candid identity service.
+Currently there is no way to specify username/password for the source charmstore.`)
+}
+
+func main() {
+	var auth authInfo
+	gnuflag.Var(&auth, "auth", "user:passwd to use for basic HTTP authentication to destination URL")
+	gnuflag.Usage = func() {
+		printCmdUsage()
+		os.Exit(2)
+	}
+
+	gnuflag.Parse(true)
+	if gnuflag.NArg() != 2 {
+		gnuflag.Usage()
+	}
+
+	fileName := gnuflag.Arg(0)
+	destURL := gnuflag.Arg(1)
+
+	whitelist, err := parseWhitelistFile(fileName)
+	if err != nil {
+		fatalf("unable to parse whitelist: %v", err)
+	}
+
+	jar, err := cookiejar.New(nil)
+	if err != nil {
+		fatalf("unable to create cookie jar: %v", err)
+	}
+	defer func() {
+		err := jar.Save()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "warning: unable to save cookie jar: %v\n", err)
+		}
+	}()
+
+	bakeryClient := httpbakery.NewClient()
+	bakeryClient.Jar = jar
+	bakeryClient.AddInteractor(httpbakery.WebBrowserInteractor{})
+
+	stats := ingest.Ingest(ingest.IngestParams{
+		Src:       newCharmStoreClient(sourceURL(), bakeryClient, nil),
+		Dest:      newCharmStoreClient(destURL, bakeryClient, &auth),
+		Whitelist: whitelist,
+	})
+
+	for _, e := range stats.Errors {
+		// TODO add a callback to IngestParams so that we can print these as they happen?
+		fmt.Fprintf(os.Stderr, "error: %v\n", e)
+	}
+
+	fmt.Printf("total %d revisions of %d entities\n", stats.EntityCount, stats.BaseEntityCount)
+	if stats.ArchivesCopiedCount > 0 {
+		fmt.Printf("copied %d revisions\n", stats.ArchivesCopiedCount)
+	} else {
+		fmt.Println("no revisions copied")
+	}
+	if stats.FailedEntityCount > 0 {
+		fmt.Printf("failed to copy %d revisions\n", stats.FailedEntityCount)
+	}
+}
+
+// parseWhitelistFile takes a file name, and parses the whitelist from that file
+// returning whitelist entities. An error is returned if a file is unable to be
+// opened from the provided path, or the file is not a valid whitelist.
+func parseWhitelistFile(fileName string) ([]ingest.WhitelistEntity, error) {
+	file, err := os.Open(fileName)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	return parseWhitelist(fileName, file)
```

Comment:
Again parseWhitelist doesn't really have a use for the file name except for filling out an error. It might be better to add the filename to the error in this function (it might also be a lot of noise for not much gain).

Reply:
Hmmm I get what you're saying. I'm tempted to leave, certainly in this PR